### DISCUSSION TITLE
UIButton Styling with Functions

### DISCUSSION THREAD

**Discussion by AndreiVidrasco** [pointfreeco/episode-code-samples#86] [Q&A]
Hi, 

I enjoyed a lot episode #3 about uikit styling and wanted to introduce this concept in my codebase. While trying to convert from inheritance to functions, I stumbled to one use case.

```
@IBDesignable final public class ActionButton: BaseButton {
    override public func commonInit() {
        super.commonInit()
        //Some styling
        addTarget(self, action: #selector(addFeedback), for: .touchDown)
    }
    
    @objc private func addFeedback() {
        let feedback = UISelectionFeedbackGenerator()
        feedback.selectionChanged()
    }
}
```

To convert styling is easy (basically everywhere where it's used ActionButton, changing to UIButton and apply specified style), but I couldn't think of a nice solution which will allow me with one function to enable haptic feedback on pressing button

Do you have any ideas how this could be implemented in a functional way?

**Comment by stephencelis** [comment]
Hi @AndreiVidrasco! Thanks for watching! Instead of subclassing, you could write a styling function that introduces a new object to hold the `addFeedback` selector, add it as a button target for that selector, and finally use the Objective-C runtime to set it as an associated object to your button (using `objc_setAssociatedObject`). Let me know if you have any questions!
